plasmapy/atomic/ionization_states.py

```diff
@@ -0,0 +1,965 @@
+        absolute tolerance of `tol` if dimensionless.  Alternatively,
+        this argument may be a `~astropy.units.Quantity` that represents
+        the number densities of each neutral/ion.
+
+    T_e: ~astropy.units.Quantity, keyword-only, optional
+        The electron temperature or thermal energy per particle.
+
+    n_e: ~astropy.units.Quantity, keyword-only, optional
+        The electron number density.
+
+    n_elem: ~astropy.units.Quantity, keyword-only, optional
+        The number density of the element, including neutrals and all
+        ions.
+
+    tol: float or int, keyword-only, optional
+        The absolute tolerance used by `~numpy.isclose` when testing
+        normalizations and making comparisons.  Defaults to `1e-15`.
+
+    Raises
+    ------
+    ~plasmapy.utils.AtomicError
+        If the ionic fractions are not normalized or contain invalid
+        values
+
+    ~plasmapy.utils.InvalidParticleError
+        If the particle is invalid.
+
+    Examples
+    --------
+    >>> states = IonizationState('H', [0.6, 0.4], n_elem=1*u.cm**-3, T_e=11000*u.K)
+    >>> states.ionic_fractions
+    array([0.6, 0.4])
+    >>> states.n_e  # electron number density
+    <Quantity 400000. 1 / m3>
+    >>> states.n_elem  # element number density
+    <Quantity 1000000. 1 / m3>
+
+    Notes
+    -----
+    Only one of `n_e` and `n_elem` may be set.
+
+    """
+
+    # TODO: Allow this class to (optionally?) handle negatively charged
+    # ions.  There are instances where singly negatively charged ions
+    # are important in astrophysical plasmas, such as H- in the
+    # atmospheres of relatively cool stars.  There may be some rare
+    # situations where doubly negatively charged ions show up too,
+    # though triply negatively charged ions are very unlikely.
+
+    # TODO: Add in functionality to find equilibrium ionization states.
+    # How much data will this require?
+
+    @check_quantity({
+        "T_e": {"units": u.K, "none_shall_pass": True},
+        "n_e": {"units": u.m ** -3, "none_shall_pass": True},
+        "n_elem": {"units": u.m ** -3, "none_shall_pass": True},
+    })
+    @particle_input(require='element', exclude='ion')
+    def __init__(self,
+                 particle: Particle,
+                 ionic_fractions=None,
+                 *,
+                 T_e=None,
+                 n_e=None,
+                 n_elem=None,
+                 tol: Union[float, int] = 1e-15):
+        """Initialize a `~plasmapy.atomic.IonizationState` instance."""
+
+        self._particle = particle
+
+        try:
+            self.tol = tol
+            self.T_e = T_e
+            self.n_elem = n_elem
+            self.n_e = n_e
+            self.ionic_fractions = ionic_fractions
+
+            # This functionality has not yet been implemented:
+
+            # if self._ionic_fractions is None and self.T_e is not None:
+            #     self.equilibrate()
```

Comment:
`raise NotImplementedError` exists for this very purpose, and it's going to show us an uncovered line!